src/frontend/src/optimizer/plan_node/logical_join.rs

```diff
@@ -1319,41 +1319,60 @@ impl ToStream for LogicalJoin {
         let r2i = join.core.r2i_col_mapping().composite(&mapping);
 
         // Add missing pk indices to the logical join
-        let left_to_add = left
+        let mut left_to_add = left
             .logical_pk()
             .iter()
             .cloned()
-            .filter(|i| l2i.try_map(*i).is_none());
+            .filter(|i| l2i.try_map(*i).is_none())
+            .collect_vec();
 
-        let right_to_add = right
+        let mut right_to_add = right
             .logical_pk()
             .iter()
             .cloned()
             .filter(|i| r2i.try_map(*i).is_none())
-            .map(|i| i + left_len);
+            .map(|i| i + left_len)
+            .collect_vec();
 
         // NOTE(st1page): add join keys in the pk_indices a work around before we really have stream
         // key.
         let right_len = right.schema().len();
         let eq_predicate = EqJoinPredicate::create(left_len, right_len, join.on().clone());
 
-        let left_to_add = left_to_add
-            .chain(
-                eq_predicate
-                    .left_eq_indexes()
-                    .into_iter()
-                    .filter(|i| l2i.try_map(*i).is_none()),
-            )
-            .unique();
-        let right_to_add = right_to_add
-            .chain(
-                eq_predicate
-                    .right_eq_indexes()
-                    .into_iter()
-                    .filter(|i| r2i.try_map(*i).is_none())
-                    .map(|i| i + left_len),
-            )
-            .unique();
+        let (need_left_jk, need_right_jk) = match self.join_type() {
+            JoinType::Inner | JoinType::LeftOuter | JoinType::LeftSemi | JoinType::LeftAnti => {
```

Comment:
I have a feeling that for `Inner`, it should be `(false, false)`?

Reply:
With this current code, for inner join, if we have right in output but not left, we will put left into output.

Reply:
Oh, I guess for inner, it we will always have left side as pk. Well, it's not ideal, but I guess our choice of PK is not output_indices aware.

Reply:
The `StreamHashJoin` derives its distribution based on the left join keys for the `Inner` join type, meanwhile, logical pk needs to comprise the distribution key, so I choose to set `need_left_jk` true for the `Inner` type.